From 40b49b2dfb3e329acf135cd6c5c7e2fa7a052fda Mon Sep 17 00:00:00 2001
From: Daniel Scally <djrscally@gmail.com>
Date: Fri, 11 Sep 2020 00:06:11 +0100
Subject: [PATCH] Fix bugs in software_node_graph_get_next_endpoint()

As originally written, the software_node_graph_get_next_endpoint() function
holds on to references for port software_nodes and additionally passes invalid
combinations of parameters to software_node_get_next_child - an endpoint which
is not a child of the port that is also passed.

This patch rectifies both issues.

Signed-off-by: Daniel Scally <djrscally@gmail.com>
---
 drivers/base/swnode.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/base/swnode.c b/drivers/base/swnode.c
index 3667467196f0..5a28408d9137 100644
--- a/drivers/base/swnode.c
+++ b/drivers/base/swnode.c
@@ -581,10 +581,17 @@ software_node_graph_get_next_endpoint(const struct fwnode_handle *fwnode,
 	}
 
 	for (; port; port = swnode_graph_find_next_port(parent, port)) {
+
+		if ((old) && (software_node_get_parent(old) != port ))
+			old = NULL;
+
 		endpoint = software_node_get_next_child(port, old);
 		fwnode_handle_put(old);
+
 		if (endpoint)
-			break;		
+			break;
+		else
+			fwnode_handle_put(port);
 	}
 
 	fwnode_handle_put(port);
-- 
2.17.1

